storage/localstore/mode_set.go

```diff
@@ -175,50 +175,61 @@ func (db *DB) set(mode chunk.ModeSet, addr chunk.Address) (err error) {
 		}
 
 	case chunk.ModeSetPin:
-
 		// Get the existing pin counter of the chunk
-		existingPinCounter, err := db.GetPinCounterOfChunk(item.Address)
+		existingPinCounter := uint64(0)
+		pinnedChunk, err := db.pinIndex.Get(item)
 		if err != nil {
 			if err == leveldb.ErrNotFound {
 				// If this Address is not present in DB, then its a new entry
 				existingPinCounter = 0
 			} else {
 				return err
 			}
+		} else {
+			existingPinCounter = pinnedChunk.PinCounter
 		}
 
-		// Otherwise increase the existng counter by 1
-		item.PinCounter = existingPinCounter + 1
+		// If this is the first time the chunk is getting pinned
+		// remove this from the gcIndex
+		if existingPinCounter == 0 {
+			db.gcIndex.DeleteInBatch(batch, item)
 
-		db.pinIndex.PutInBatch(batch, item)
+			// a check is needed for decrementing gcSize
+			// as delete is not reporting if the key/value pair
+			// is deleted or not
+			if _, err := db.gcIndex.Get(item); err == nil {
+				gcSizeChange = -1
+			}
+		}
 
+		// Otherwise increase the existing counter by 1
+		item.PinCounter = existingPinCounter + 1
+		db.pinIndex.PutInBatch(batch, item)
 	case chunk.ModeSetUnpin:
 		// Get the existing pin counter of the chunk
-		existingPinCounter, err := db.GetPinCounterOfChunk(item.Address)
+		pinnedChunk, err := db.pinIndex.Get(item)
 		if err != nil {
 			return err
 		}
 
 		// Decrement the pin counter or
-		// delete it from pin index if the oin counter has reached 0
-		if existingPinCounter > 1 {
-			item.PinCounter = existingPinCounter - 1
+		// delete it from pin index if the pin counter has reached 0
+		if pinnedChunk.PinCounter > 1 {
+			item.PinCounter = pinnedChunk.PinCounter - 1
 			db.pinIndex.PutInBatch(batch, item)
 		} else {
 			db.pinIndex.DeleteInBatch(batch, item)
-		}
-
-	case chunk.ModeSetRawFile:
-		item.IsRaw = 1
-		db.pinFilesIndex.PutInBatch(batch, item)
 
-	case chunk.ModeSetFile:
-		item.IsRaw = 0
-		db.pinFilesIndex.PutInBatch(batch, item)
-
-	case chunk.ModeSetUnpinFile:
-		db.pinFilesIndex.DeleteInBatch(batch, item)
+			// If the file is getting unpinned, add it to gcIndex
+			db.gcIndex.PutInBatch(batch, item)
 
+			// a check is needed for decrementing gcSize
```

Comment:
Yes, actually, that is completely true, but only if mode set sync is checking if the item is not pinned when adding to the gc index. And it looks like that this is missing. So, two problems here:

 - gcSize needs to be incremented when adding item to the gc index
 - the check in ModeSetSync if the chunk is pinned is needed before adding to the gc index, and then we can assume here that the chunk is not in the gc index and remove the Get call

Am I missing something else, maybe?

Reply:
Added the second point that @janos suggested. I am now checking in ModeSetSync if the chunk is in pinIndex and adding to GC only if it is not pinned. Because of this i am removing the GC fiddling here.